precommit/src/main/shell/test-patch-docker/Dockerfile

```diff
@@ -145,14 +145,22 @@ RUN curl -sSL \
 ####
 # Install hadolint (dockerfile lint)
 ####
-FROM yetusbase AS yetushadolint
+FROM yetusbase as yetushadolint_arm64
 ARG HADOLINT_VERSION=2.12.0
 SHELL ["/bin/bash", "-o", "pipefail", "-c"]
-RUN if [[ "$(uname -m)" == "x86_64" ]]; then curl -sSL \
-        https://github.com/hadolint/hadolint/releases/download/v$HADOLINT_VERSION/hadolint-Linux-"$(uname -m)" \
+RUN curl -sSl \
```

Comment:
Can we use `COPY/ADD --link` in place of this embedded curl command? I think it caches better, but requires `docker/dockerfile:1.4` or later. https://docs.docker.com/engine/reference/builder/#copy---link

Reply:
Probably worth experimenting with now. 

There were a few evolutions.  `ADD` + `RUN` would generate 2 layers instead of a single `RUN` layer.  Those extra layers added up in the final image making it much bigger than necessary. When we only built images for x86, there were checksums for all of the files.  With multi-arch + multi-stage builds, those two problems are mostly non-existent now.

Caching should be pretty good withe multi-stage builds so long as `yetusbase` doesn't change.  If it does change, in some cases one really wants to rebuild that stage anyway.  But that wouldn't have an impact on `ADD` anyway.

Reply:
Also, why not just add an elif or whatever statement here rather do two downloads and have to figure out which hadolint should be in place later on in the Dockerfile?

Reply:
Docker's runtime prunes the execution graph such that the download only happens for the architecture target that is being built. If you're building a multi-articture image, both binaries are downloaded and both are used by independent copies of the subsequent step. That step is only defined once (keeping things DRY) and the downloaded binaries are both cached, which eases developer pain. I agree that it adds to layer accumulation.